go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -0,0 +1,1124 @@
+		Reconcile:      c.reconcileEgressIP,
+		Threadiness:    1,
+		Informer:       wf.EgressIPInformer().Informer(),
+		Lister:         wf.EgressIPInformer().Lister().List,
+		ObjNeedsUpdate: egressIPNeedsUpdate,
+		HandleError:    handleError,
+	}
+	c.eipController = controllerutil.NewController("clustermanager routeadvertisements egressip controller", eipConfig)
+
+	return c
+}
+
+func (c *Controller) Start() error {
+	defer klog.Infof("Cluster manager routeadvertisements started")
+	return controllerutil.Start(
+		c.eipController,
+		c.frrController,
+		c.nadController,
+		c.nodeController,
+		c.raController,
+	)
+}
+
+func (c *Controller) Stop() {
+	controllerutil.Stop(
+		c.eipController,
+		c.frrController,
+		c.nadController,
+		c.nodeController,
+		c.raController,
+	)
+	klog.Infof("Cluster manager routeadvertisements stoppedu")
+}
+
+func (c *Controller) ReconcileNetwork(name string, old, new util.NetInfo) {
+	// this controller already listens on NAD events however we skip NADs
+	// pointing to networks that network manager is still not aware of; so we
+	// only need to signal the reconciliation on new networks and it is enough
+	// to do so with a single of its NADs
+	if new == nil || old != nil {
+		return
+	}
+	c.nadController.Reconcile(new.GetNADs()[0])
+}
+
+// Reconcile RouteAdvertisements. For each selected FRRConfiguration and node,
+// another FRRConfiguration might be generated:
+//
+// - If pod network advertisements are enabled, the generated FRRConfiguration
+// will announce from the node the selected network prefixes for that node on
+// the matching target VRFs.
+//
+// - If EgressIP advertisements are enabled, the generated FRRConfiguration will
+// announce from the node the EgressIPs allocated to it on the matching target
+// VRFs.
+//
+// - If pod network advertisements are enabled, the generated FRRConfiguration
+// will import the target VRFs on the selected networks as required.
+//
+// - The generated FRRConfiguration will be labeled with the RouteAdvertisements
+// name and annotated with an internal key to facilitate updating it when
+// needed.
+//
+// The controller will also annotate the NADs of the selected networks with the
+// RouteAdvertisements that select them to facilitate processing for downstream
+// zone/node controllers.
+//
+// Finally, it will update the status of the RouteAdvertisements.
+//
+// The controller processes selected events of RouteAdvertisements,
+// FRRConfigurations, Nodes, EgressIPs and NADs.
+func (c *Controller) reconcile(name string) error {
+	startTime := time.Now()
+	klog.V(5).Infof("Syncing routeadvertisements %q", name)
+	defer func() {
+		klog.V(4).Infof("Finished syncing routeadvertisements %q, took %v", name, time.Since(startTime))
+	}()
+
+	ra, err := c.raLister.Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get RouteAdvertisements %q: %w", name, err)
+	}
+
```

Comment:
if ra is nil cause it was not found, you still want to continue to reconcileRouteAdvertisements right? just making sure...

Reply:
yeah, that means it is a delete and I need to clean up, reconcile will do that.